device/device.go

```diff
@@ -102,40 +126,54 @@ func (d *Device) Connect(result chan<- error) {
 			return
 		}
 
-		// First of all, ensure we have a certificate
-		if err := d.ensureCertificate(); err != nil {
+		// Define a retry policy and the operation to be executed, i.e. we want to
+		// retrieve the brokerURL with an HTTP request
+		policy := d.makeRetryPolicy()
+		operation := func() error {
+			err := d.getBrokerURL()
+
+			if err != nil && d.ConnectRetry {
+				fmt.Println("Cannot retrieve brokerURL. Retrying...")
+			}
+			return err
+		}
+
+		err := backoff.Retry(operation, policy)
+		if err != nil && d.ConnectRetry {
 			if result != nil {
-				result <- err
+				errorMsg := fmt.Sprintf("Cannot establish a connection after %d attempts.", d.connectionRetryAttempts())
+				result <- errors.New(errorMsg)
 			}
 			return
 		}
 
-		brokerURL, err := d.getBrokerURL()
-		if err != nil {
+		if err != nil && !d.ConnectRetry {
 			if result != nil {
 				result <- err
 			}
 			return
 		}
 
-		if err := d.initializeMQTTClient(brokerURL); err != nil {
+		// Ensure we have a certificate
+		if err := d.ensureCertificate(); err != nil {
 			if result != nil {
 				result <- err
 			}
 			return
 		}
```

Comment:
You also have to wrap this in a retry operation. Otherwise if you lose connection after getBrokerUrl but before calling this, you don't honor the retry mechanism.